### PR TITLE
Switch to STM-based Behaviors

### PR DESCRIPTION
**Pull request by bts**
This affords atomicity/consistency in the face of multiple producers to a `Behavior`.

### CODE REVIEW

File: src/QuorumTools/Control.hs [hunk 1]
```diff
@@ -52,35 +47,33 @@ find' predicate = Fold step Nothing id where
...
-observe (Behavior _ mvar) = liftIO $ readMVar mvar
+observe (Behavior _ mvar) = liftIO $ atomically $ readTMVar mvar
 
 --
 -- TODO: this won't leak memory via the duplicated channel after our event
 -- occurs, right?
 --
 watch :: MonadManaged m => Behavior a -> (a -> Maybe b) -> m (Async b)
-watch b check = do
+watch b decide = do
```
Comment (bts) [inline src/QuorumTools/Control.hs]: This is because there's an existing STM function called `check`.

File: src/QuorumTools/Control.hs [hunk 2]
```diff
@@ -52,35 +47,33 @@ find' predicate = Fold step Nothing id where
...
   chan <- subscribe b
-  fork $ liftIO $ untilJust $ check <$> readChan chan
+  fork $ liftIO $ untilJust $ decide <$> atomically (readTChan chan)
```
Comment (joelburget) [inline src/QuorumTools/Control.hs]: Oh that's just how you embed it in `IO`.

Comment (bts) [inline src/QuorumTools/Control.hs]: Yeah, it's necessary -- without `atomically`, you only have `STM a`. `atomically` runs the `STM a` action to produce an `IO a`.

File: src/QuorumTools/Control.hs [hunk 3]
```diff
@@ -52,35 +47,33 @@ find' predicate = Fold step Nothing id where
...
-subscribe (Behavior chan _) = liftIO $ dupChan chan
+subscribe :: MonadIO m => Behavior a -> m (TChan a)
+subscribe (Behavior chan _) = liftIO $ atomically $ dupTChan chan
 
 observe :: MonadIO m => Behavior a -> m a
-observe (Behavior _ mvar) = liftIO $ readMVar mvar
+observe (Behavior _ mvar) = liftIO $ atomically $ readTMVar mvar
 
 --
 -- TODO: this won't leak memory via the duplicated channel after our event
```
Comment (joelburget) [inline src/QuorumTools/Control.hs]: I'm not sure I understand the question. Is this still a concern?

**Pull request merged**